src/cockpit/channels/metrics.py

```diff
@@ -15,23 +15,179 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+import asyncio
+import json
+import sys
+import time
 import logging
+from typing import Any, Dict, List, Set, NamedTuple, Optional
+from collections import defaultdict
 
-from ..channel import Channel
+from ..channel import AsyncChannel
+from ..samples import SAMPLERS, BaseSampler, SampleDescription
 
 logger = logging.getLogger(__name__)
 
 
-class MetricsChannel(Channel):
+class MetricInfo(NamedTuple):
+    name: str
+    derive: Optional[str]
+    units: str
+    semantics: str
+    instanced: bool
+
+
+class InternalMetricsChannel(AsyncChannel):
     payload = 'metrics1'
+    restrictions = {'source': 'internal'}
+
+    metrics: List[MetricInfo]
+    samplers: Set
+
+    samples: Dict[str, Any]
+    old_samples: Dict[str, Any]
+
+    samplers_cache: Optional[Dict[str, BaseSampler]] = None
+    description_cache: Optional[Dict[str, SampleDescription]] = None
+
+    interval: int = 1000
+    need_meta: bool = False
+    last_timestamp: int = 0
+    next_timestamp: int = 0
+
+    @classmethod
+    def ensure_samplers(cls):
+        if cls.samplers_cache is None:
+            cls.samplers_cache = {}
+            cls.descriptions_cache = {}
+            for sampler in SAMPLERS:
+                print(sampler)
+                for name, desc in sampler.descriptions.items():
+                    cls.samplers_cache[name] = sampler
+                    cls.descriptions_cache[name] = desc
+
+    def parse_options(self, options):
+        sampler_classes = set()
+        logger.debug("metrics internal open: %s, channel: %s", options, self.channel)
+
+        interval = options.get('interval', self.interval)
+        if interval <= 0 or interval > sys.maxsize:
+            logger.error('invalid "interval" value: %s', interval)
+            self.close(problem="protocol-error", message='invalid interval option')
+            return False
+
+        self.interval = interval
+
+        metrics = options.get('metrics')
+        if not metrics or not isinstance(metrics, list):
```

Comment:
What is the logic that you're trying to implement here?  Because I read this backwards from what you probably intend.  Specifically:

 - I think you're trying to check for `None` left of the `and`, but you're also rejecting empty lists there (ie: rewrite to `is None`).
 - If it's really specifically `None` that you want to check, that could be handled with the `isinstance` check, since `None` is obviously not a `list`.

ie: if your intention is to check that it's a list *and* not empty, I'd rather more intuitively write:

`if not instance(metrics, list) or len(metrics) == 0:` or so